### PR TITLE
Add Voting.sol

### LINKED ISSUE #92: Voting and Staking
Consider how a voting library could be built for DAO.sol.

The idea is that there are several actions that can be voted. The actions are known at coding time (for example, investing in a new venture, distributing dividends and raising funds).

Voting is the same for all actions. There is a staking token and each token is one vote. Each token cannot be used for more than one voting simultaneously.

I would like the logic to be reused between voting actions, only changing the resulting action if the vote is successful.

We might have to mess with function signatures and things like that, no idea really.

### PR DESCRIPTION
**Pull request by uivlis**
Fixes #92. 
I thought that you don't need both "yes" and "no" votes, just "yes" since if you want to vote no, you can simply not vote. And because of this, I also dropped the "majority" param, keeping instead just the "quorum". If the quorum (as a percentage of total supply) is met, then the voting proposal is enacted. If not, the voting will be canceled.
I know I should have added tests, but I wanted to check first if this is the right way to go. I will add them to this PR if all is well.

### CODE REVIEW

File: contracts/drafts/voting/Voting.sol [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    uint256 public quorum;
+
+    /**
+     * @dev Initialize the issuance with the token to issue and the token to
+     * accept as payment.
+     */
+    constructor(
+        address _votingToken,
+        address _proposalContract,
+        bytes memory _proposalData
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: That's cool, I didn't know how to code the arguments of the proposal into a single variable :+1:

File: contracts/drafts/voting/Voting.sol [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        _createTransition("SETUP", "OPEN");
+        _createTransition("OPEN", "PASSED");
+        _createTransition("OPEN", "FAILED");
+        emit VotingCreated();
+    }
+
+    /**
+     * @dev Function to enact the proposal of this voting
+     */
+    function enact() external payable onlyOwner {
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: Let's remove `onlyOwer` from here, it means giving a baked in veto to the address deploying the contract, which might not always be right.

File: contracts/drafts/voting/Voting.sol [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        require(
+            currentState == "PASSED",
+            "Cannot enact proposal now."
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: "Cannot enact proposal until vote passes."

File: contracts/drafts/voting/Voting.sol [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        );
+        // solium-disable-next-line security/no-call-value
+        (bool success, ) = proposalContract
+            .call.value(msg.value)(proposalData);
+        require(success, "Could not enact proposal.");
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: "Failed to enact proposal."

File: contracts/drafts/voting/Voting.sol [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        }
+        votes[msg.sender] = votes[msg.sender].add(_votes);
+        emit VoteCasted(msg.sender, _votes);
+    }
+
+    /**
+     * @dev Use this function to retrieve your votingToken votes in case you changed your mind.
+     */
+    function cancel() external {
+        require(
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: Do we need to forbid this?

File: contracts/drafts/voting/Voting.sol [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        IERC20(votingToken).transfer(msg.sender, count);
+        emit VoteCancelled(msg.sender, count);
+    }
+
+    /**
+     * @dev Function to open the voting
+     */
+    function open() public onlyOwner {
+        require(
+            quorum > 0,
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: Please rename "[quorum](https://en.wikipedia.org/wiki/Quorum)" to "[threshold](https://en.wikipedia.org/wiki/Electoral_threshold)" throughout.

File: contracts/drafts/voting/Voting.sol [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * @dev Function to validate the quorum
+     */
+    function validate() public onlyOwner {
+        ERC20Detailed _votingToken = ERC20Detailed(votingToken);
+        int256 totalSupplyFixed = _votingToken.totalSupply()
+            .safeUintToInt().newFixed(_votingToken.decimals());
+        int256 quorumFixed = quorum.safeUintToInt().newFixed(2);
+        uint256 threshold = totalSupplyFixed.multiply(quorumFixed)
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: Mmm, if we use `threshold` already, I guess this should be `yesPercentage`.

File: contracts/drafts/voting/Voting.sol [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            .fromFixed(_votingToken.decimals()).safeIntToUint();
+        require(
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: Please move the code above this comment to a function that returns the percentage of yes votes.

File: contracts/drafts/voting/Voting.sol [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    function cancelAllVotes() public onlyOwner {
+        _transition("FAILED");
+    }
+
+    /**
+     * @notice Set the quorum during SETUP state. Quorum must be a percentage of the votingToken's totalSupply, with the comma shifted two digits to the right. For example:
+     * If desired quorum is 100% (unanimity), then pass to this function: 10000.
+     * If desired quorum is 66.67% (super majority), then passs to this function: 6667.
+     */
+    function setQuorum(uint256 _quorum) public onlyOwner {
```
Comment (albertocuestacanada) [inline contracts/drafts/voting/Voting.sol]: Why don't we just pass the quorum / threshold as a parameter to the constructor?

**Comment by uivlis** [comment]
Yes, that is true, however, isn't it repetitive to have both `quorum` and `threshold`? You could have `justThreshold` = `threshold` / 100 * `quorum` and the resulting vote would be equivalent:
```
1. threshold and quorum => votes pass if total > threshold / 100 * quorum / 100 * totalSupply
2. justThreshold => votes pass if total > justThreshold / 100 * totalSupply
```

**Comment by albertocuestacanada** [comment]
> A [quorum](https://en.wikipedia.org/wiki/Quorum) is the minimum number of members of a deliberative assembly (a body that uses parliamentary procedure, such as a legislature) necessary to conduct the business of that group. According to Robert's Rules of Order Newly Revised, the "requirement for a quorum is protection against totally unrepresentative action in the name of the body by an unduly small number of persons."

**Comment by albertocuestacanada** [comment]
You could argue that it is not always necessary, and I'm not going to push for it now, it can be added on later.

**Comment by uivlis** [comment]
I get that, but that requirement could still be met by the current arrangement. If both quorum and threshold are expressed in percentages, you can merge them. For example:
1. Quorum = 50%, threshold = 50% => just_one_threshold = 25%
1. quorum = 100%, threshold = 66.67% => just_one_threshold = 66.67%

**Comment by albertocuestacanada** [comment]
Let's drop the very interesting politics theory discussion. Please implement a single threshold and when people ask us to also implement the quorum we'll do that, power to the people. :man_shrugging:

**Comment by uivlis** [comment]
Right, so if this is good enough, should I begin testing?

**Comment by albertocuestacanada** [comment]
Yeah, go for it :+1:

**Comment by uivlis** [comment]
After testing, I decided to change the proposal mechanism, so that you can register more proposals before opening the voting and, after the validation, enact them one by one. This helps in the case of Issuances (or Ventures) in order to be able to claim those tokens, otherwise they would be lost forever.

**Comment by albertocuestacanada** [comment]
@uivlis, I removed the `onlyOwner` modifier from `validate` since it is a veto for the contract owner, which can choose to never validate a vote if he doesn't like the outcome.

For the same reason, I removed the FAILED state.

The result is a bit of a weird vote, where all votes are always open. I would think that in a production implementation this could be managed by allowing anyone to cancel a vote if it has less than some defined percentage of votes. I don't think that we need to worry too much.

**Comment by albertocuestacanada** [comment]
Please have a look and merge if you agree with the changes. Be careful because my latest changes in `StateMachine` will break `Voting`. To fix you just need to remove the `_createState` calls.

**Pull request merged**